### PR TITLE
Add prefix argument to mktempdir()

### LINKED ISSUE #22922: mktempdir() should provide an option of a prefix very similar to you can do with mkdtemp in posix
I think it may help if `mktempdir()` can provide an option of a prefix very similar to you can do with `mkdtemp` in POSIX. This provides the module developers to create temp directories with their own prefix which can help them able to track the directories created by them than everything starting with `tmp` or `jl_` prefix.

The discussion thread on this topic can be seen here: https://discourse.julialang.org/t/mktempdir-should-provide-scope-to-declare-a-prefix/5020

### PR DESCRIPTION
**Pull request by sambitdash**
Continuation of #23237, which has all review comments (remerging and rebasing detached the old PR and I cannot reopen it).

Fixes  #22922

### CODE REVIEW

**Comment by sambitdash** [comment]
@nalimilan, @fredrikekre, @vtjnash, @kshyatt, Please review #23237 here. 

@nalimilan Comments on the test cases are incorporated. 

On the `base/file.jl` waiting on the further comments on approach for `readdir` vs. `readlink` way of handling the `req` variable in `libuv`.

File: base/file.jl
Comment (sambitdash) [inline base/file.jl]: Decision to be taken if `readdir` or `readlink` approach be taken in allocating this variable. @vtjnash  to recommend.

File: base/file.jl
Comment (sambitdash) [inline base/file.jl]: Formatting to be fixed.

File: test/file.jl
Comment (nalimilan) [inline test/file.jl]: `/` does *not* seem to be accepted, according to the test.

File: test/file.jl
Comment (nalimilan) [inline test/file.jl]: Better also test the path before the base name to ensure that what the comment says actually happens. Same below and on Unix.

**Review by nalimilan — COMMENTED** [review]
Thanks, should be good after the few test fixes and validation by @vtjnash.

File: base/file.jl
Comment (JeffBezanson) [inline base/file.jl]: `"$temp_prefix"` == `temp_prefix`

Comment (nalimilan) [reply base/file.jl]: The idea is to use the value of the global variable rather than repeating it. Seems to work as expected:
```julia
julia> x = 1
1

julia> """x = "$x" """
       x
x

help?> x

  x = "1" 
```

File: test/file.jl
Comment (sambitdash) [reply test/file.jl]: `/` only as the first character of a prefix creates problems. If a path is formed by `$parent/someval/` where a path may exist in the system like say /tmp/. It will work. So technically, '/' is supported as a prefix character if the final path is writable by the process. 

"""
julia> mktempdir("/";prefix="tmp/")
"/tmp/9dLil5"

julia> 

julia> mktempdir("/";prefix="tmp/s")
"/tmp/sZZDLWl"

julia> mktempdir(;prefix="/")
ERROR: mktempdir: permission denied (EACCES)
Stacktrace:
 [1] uv_error at ./libuv.jl:68 [inlined]
 [2] #mktempdir#12(::String, ::Function, ::String) at ./file.jl:359
 [3] (::getfield(Base.Filesystem, Symbol("#kw##mktempdir")))(::Array{Any,1}, ::typeof(mktempdir), ::String) at ./<missing>:0 (repeats 2 times)

julia> mktempdir(;prefix="ss/")
ERROR: mktempdir: no such file or directory (ENOENT)
Stacktrace:
 [1] uv_error at ./libuv.jl:68 [inlined]
 [2] #mktempdir#12(::String, ::Function, ::String) at ./file.jl:359
 [3] (::getfield(Base.Filesystem, Symbol("#kw##mktempdir")))(::Array{Any,1}, ::typeof(mktempdir), ::String) at ./<missing>:0 (repeats 2 times)

julia> mkdir("/tmp/ss");mktempdir(;prefix="ss/")
"/tmp/ss/ZDAbPm"

julia> 
"""

Comment (nalimilan) [reply test/file.jl]: OK, so just make the comment more explicit.

File: test/file.jl
Comment (nalimilan) [inline test/file.jl]: In most cases the user running Julia won't have access to that path (fortunately), so we'll have to find something else. Maybe pass `tempdir()` as a prefix, since we know we can write there?

File: test/file.jl
Comment (nalimilan) [inline test/file.jl]: Would be nice to also remove the directory in other cases.

File: test/file.jl
Comment (sambitdash) [reply test/file.jl]: Do you expect me to test for the following?
```
mktempdir("c:/mydir"; prefix="c:/") do tmpdir
   @test isdir(tmpdir)  
          # Does not makes much of sense as mktempdir() will create the dir() anyway. 
          # Our focus is mostly the name of the dir created. 
          # Those must be tested as part of regular mktempdir (without prefix) test case. 
   filename = basename(tmpdir)
```
The file already has a large number of `mktempdir(without prefix)` test cases.

File: base/file.jl
Comment (JeffBezanson) [reply base/file.jl]: My mistake --- I didn't notice this was in a doc string :)

File: test/file.jl
Comment (nalimilan) [reply test/file.jl]: I meant that since you say "The temp directory created is of form "C:\\XXXXXX"", you should actually test that it's the case, i.e. that `startswith(tempdir, C:\\")`.

**Comment by vtjnash** [comment]
> On the base/file.jl waiting on the further comments on approach for readdir vs. readlink way of handling the req variable in libuv.

Yikes! That's quite a mess of different approaches in there. OTOH, you can use either approach here.

**Comment by sambitdash** [comment]
Considering the comment from @vtjnash, I have committed only the formatting changes needed in the `base/file.jl`. That should address all the comments so far received on the PR.

**Review by nalimilan — APPROVED** [review]
Looks good to me, as long as @vtjnash agrees.

**Comment by sambitdash** [comment]
The errors on CircleCI do not seem relevant to the PR and Travis is stuck for some reason.

**Comment by sambitdash** [comment]
Both the Travis and CircleCI seem to report spurious errors.

**Comment by sambitdash** [comment]
closing and reopening to rerun tests.

**Comment by sambitdash** [comment]
Errors reported by Travis and CircleCI seem to be unrelated to the check-in. In fact in the earlier iterations the 64-bit version on Travis was passing while the Mac OS X tests didn't even kick off. Please review the changes for merge if any changes are needed let me know.

**Comment by nalimilan** [comment]
@vtjnash Merge?

File: base/file.jl
Comment (vtjnash) [inline base/file.jl]: The wrapping seems odd here. If you wrap once at the period, I think the rest will fit on the following line.

File: base/file.jl
Comment (vtjnash) [inline base/file.jl]: I think the intention of this API was to compute simply:
`tpath = "$(parent)$(path_separator)$(template)XXXXXX"`

**Comment by vtjnash** [comment]
Sorry for all of the incremental review. I know that's really annoying. But should be almost good to go :)

File: base/file.jl
Comment (nalimilan) [reply base/file.jl]: You mean that it's simpler/more efficient to allocate the string in a single pass?

Comment (vtjnash) [reply base/file.jl]: That doesn't matter here. However, the goal of the API appears to be to compute the simpler operation, not joinpath.

Comment (sambitdash) [reply base/file.jl]: `joinpath` has 2 functions:

1. Use the relevant $(path_separator)
2. Provide a sane path relevant to the OS by stripping end separator. 
```
     ismatch(path_separator_re, a[end:end]) ? string(C,a,b) :
                                             string(C,a,pathsep(a,b),b)
```

If the input to `parent` variable ends with a `$(path_separator)` then the path created will have 2 consecutive `$path_separator` values. It does not fail but is not the correct return value. 

Example:
```
julia> mktempdir("/tmp/"; prefix="abc")
"/tmp//abclEMi2T"

julia> mktempdir("/"; prefix="tmp/")
"//tmp/rXRggu"
```

File: base/file.jl
Comment (sambitdash) [reply base/file.jl]: removed in latest check-in.

File: base/file.jl
Comment (sambitdash) [reply base/file.jl]: Fixed in latest check-in.

File: base/file.jl
Comment (sambitdash) [reply base/file.jl]: @vtjnash Can you review and suggest if `joinpath` should be removed from the code.

Comment (vtjnash) [reply base/file.jl]: Yes, it should. But to your point, yes we can add some logic here to avoid duplicating the path separator if already present at the end of parent:
```
if ismatch(path_separator_re, parent[end:end])
    tpath = "$(parent)$(template)XXXXXX"
else
    tpath = "$(parent)$(path_separator)$(template)XXXXXX"
end
```

File: base/file.jl
Comment (StefanKarpinski) [reply base/file.jl]: Technically `$(repr(temp_prefix))` is more correct in case there are weird characters in the prefix (seems unlikely, but it is possible).

**Comment by sambitdash** [comment]
@StefanKarpinski, @vtjnash, 

There has been almost a few months we have been working on this PR. And I see very sporadic reviews and review fixes invariably gets into merge conflicts and rebase resolutions that unnecessarily complicate the check-ins. If we do not see much priority in this feature, may be we should discard this feature from the product. I think I have checked-in the last functional code at least 10 days back yet no further update.

**Comment by StefanKarpinski** [comment]
It's not that we don't want this functionality, but that this needs careful review before merging and people are busy – we're trying to get Julia 1.0 ready very soon. Adding features can always be done in the future. Is there some part of this PR that breaks existing functionality?

**Comment by nalimilan** [comment]
AFAICT this should be ready to go, the remaining discussions where mostly about the best coding patterns.

**Comment by StefanKarpinski** [comment]
@nalimilan, could you review it, fix the conflicts and merge if it's good to go?

**Comment by nalimilan** [comment]
I've approved it already, @vtjnash is the one who had comments.

**Comment by StefanKarpinski** [comment]
@vtjnash, please approve if you feel that your comments have been addressed.

**Comment by ViralBShah** [comment]
Any reason still pending?

File: base/file.jl
Comment (JeffBezanson) [inline base/file.jl]: I think this will include two sets of quotes (quotes are included in the output of `repr` of a string).

**Comment by StefanKarpinski** [comment]
This is non-breaking and still not correct. Removing the triage label.

File: base/file.jl
Comment (StefanKarpinski) [inline base/file.jl]: If `prefix` contains the letter `X` this is going to be wrong; it's hard to say if that's avoidable or not, given the way the underlying `uv_fs_mkdtemp` API works.

Comment (nalimilan) [reply base/file.jl]: Yes. It's not even clear why these chars need to be `X` rather than any other ASCII char.

File: base/file.jl
Comment (nalimilan) [reply base/file.jl]: See previous comment: https://github.com/JuliaLang/julia/pull/23634#discussion_r142481786

File: base/file.jl
Comment (sambitdash) [reply base/file.jl]: Links to documentation for `uv_fs_mkdtemp` and `mkdtemp` given below:

http://docs.libuv.org/en/v1.x/fs.html
https://linux.die.net/man/3/mkdtemp

**Comment by tpapp** [comment]
It would be great to revive this PR, as the feature is quite convenient.

**Pull request closed without merging**
